src/System.Reflection.MetadataLoadContext/src/System/Reflection/PathAssemblyResolver.cs

```diff
@@ -0,0 +1,117 @@
+        private static readonly byte[] s_EmptyArray = Array.Empty<byte>();
+
+        private ConcurrentDictionary<string, List<string>> _fileToPaths = new ConcurrentDictionary<string, List<string>>(StringComparer.OrdinalIgnoreCase);
+
+        /// <summary>
+        /// Initializes a new instance of the cref="System.Reflection.PathAssemblyResolver"> class.
+        /// </summary>
+        /// <exception cref="System.ArgumentNullException">Thrown when assemblyPaths is null.</exception>
+        /// <exception cref="System.ArgumentException">Thrown when a path is invalid or an Assembly simple name was specified more than once.</exception>
+        public PathAssemblyResolver(IEnumerable<string> assemblyPaths)
+        {
+            if (assemblyPaths == null)
+                throw new ArgumentNullException(nameof(assemblyPaths));
+
+            foreach (string path in assemblyPaths)
+            {
+                if (string.IsNullOrEmpty(path))
+                    throw new ArgumentException(SR.Format(SR.Arg_InvalidPath, path), nameof(assemblyPaths));
+
+                string file = Path.GetFileNameWithoutExtension(path);
+                if (file.Length == 0)
+                    throw new ArgumentException(SR.Format(SR.Arg_InvalidPath, path), nameof(assemblyPaths));
+
+                List<string> paths;
+                if (!_fileToPaths.TryGetValue(file, out paths))
+                {
+                    _fileToPaths.TryAdd(file, paths = new List<string>());
+                }
+                paths.Add(path);
+            }
+        }
+
+        public override Assembly Resolve(MetadataLoadContext context, AssemblyName assemblyName)
+        {
+            // On the first access to this assembly name, load each assembly from the provided path(s).
+            LoadMatchingAssembliesFromPaths(context, assemblyName.Name);
+
+            Assembly candidate = null;
+
+            foreach (Assembly a in context.GetAssemblies())
+            {
+                if (assemblyName.Name.Equals(a.GetName().Name, StringComparison.OrdinalIgnoreCase) && IsPublicKeyTokenCompatible(assemblyName, a))
+                {
+                    // Pick the highest version.
+                    if (candidate == null || NormalizeVersion(a.GetName().Version) > NormalizeVersion(candidate.GetName().Version))
+                    {
+                        candidate = a;
+                    }
+                }
+            }
+
+            return candidate;
+        }
+
+        private void LoadMatchingAssembliesFromPaths(MetadataLoadContext context, string simpleName)
+        {
+            List<string> paths;
+            if (_fileToPaths.TryGetValue(simpleName, out paths))
+            {
+                foreach (string path in paths)
+                {
+                    context.LoadFromAssemblyPath(path);
+                }
+
+                // Remove the path so we don't process it again.
```

Comment:
Is this really needed?

Reply:
Yes otherwise we'll try to load the same assembly(s) again later when a different AssemblyName is passed in assuming something changed like Version or PublicKeyToken from the previous load(s).

Reply:
What prevents multiple threads going through this path in parallel and hitting the problem you just described?

(I assume that the MetadataLoadContext / PathAssemblyResolver are safe for concurrent use from multiple threads.)

Reply:
There is no harm in loading the same assembly other than perf, so this avoids a possible perf hit. A multi-threaded case may end up loading the same assembly again.

Reply:
The resolver will ensure the same assembly is returned for a given Name + PublicKeyToken.

Reply:
Does this remove break the case where the same resolver is re-used for multiple contexts?

Reply:
Yes it would. We should decide on a policy - I don't think re-use is important. If we end up caching all assemblies in the resolver that would also break cases.

Reply:
The way we defined the API, there's no indication that this wouldn't be supported. If we want 1 resolver:1 context, then I think we should have an abstract context where you override resolution.

> If we end up caching all assemblies in the resolver that would also break cases.

Why would we need to cache System.Reflection.Assembly instances? Can we not just rely on loading the same path twice in the same context will return the same Assembly instance?

Reply:
Depending on what the resolver wants to do, it may need to cache, especially if we decide we don't want to loop through the previously loaded assemblies (for perf) but still want to support LoadByStream etc being used with name resolution.

So yes maybe we should change the API.

Reply:
The original proposal included a func callback which did not have this issue. However that didn't work well for code re-use \ built-in policy cases thus the separate class.

Ultimately, someone could layer on top of a func-based MetadataLoadContext and do whatever they want regarding caching or exposing only certain Load methods.

Reply:
> Can we not rely on loading the same path twice in the same context will return the same Assembly instance?

Yes you can rely on that. There are two "caches" in MetadataLoadContext:
_loadedAssemblies: All canonical assemblies that were loaded. Keyed on "name" (Name, Version, PublicKeyToken, Culture). When LoadFromStream\Path\Byte is called, an existing assembly is returned and newly read one is ignored if "name" matches 100%. If "name" matches but the Mvid is different then an exception is thrown.

_binds: All mapped assemblies populated by Resolve(). Also keyed on "name". Prevents Resolve() from being called unnecessary.

Reply:
OK based on offline mtg we agreed to above.

Reply:
The latest commit has these semantics.

Also firmed up rules about matching Assembly.PublicKeyToken depending on incoming AssemblyName.PublicKeyToken.

Also I don't believe AssemblyName.Version can ever be null when the AssemblyName is obtained from an existing assembly, so I removed the NormalizeVersion method.